test/test_qartod.py

```diff
@@ -0,0 +1,285 @@
+
+        cls.qartod_find_url = '/'.join((qartodTestServiceAPI.base_url, 'find'))
+
+    @staticmethod
+    def get_modified_ctdbp_record(**kwargs):
+        record = deepcopy(CTDBP_QARTOD_RECORD_3)
+        record.__dict__.update(kwargs)
+        return record
+
+    def create_nut_sr(self):
+        nutnr_fn = 'nutnr_a_sample.nc'
+        ctdpf_fn = 'ctdpf_sbe43_sample.nc'
+        qc = json.load(open(os.path.join(DATA_DIR, 'qc.json')))
+
+        tr = TimeRange(3.65342400e+09, 3.65351040e+09)
+        sr = StreamRequest(self.nut_sk, [18], tr, {}, qc_parameters=qc, request_id='UNIT')
+        nutnr_ds = xr.open_dataset(os.path.join(DATA_DIR, nutnr_fn), decode_times=False)
+        ctdpf_ds = xr.open_dataset(os.path.join(DATA_DIR, ctdpf_fn), decode_times=False)
+
+        nutnr_ds = nutnr_ds[self.base_params + [p.name for p in sr.stream_parameters[self.nut_sk]]]
+        ctdpf_ds = ctdpf_ds[self.base_params + [p.name for p in sr.stream_parameters[self.ctd_sk]]]
+
+        sr.datasets[self.ctd_sk] = StreamDataset(self.ctd_sk, sr.uflags, [self.nut_sk], sr.request_id)
+        sr.datasets[self.nut_sk] = StreamDataset(self.nut_sk, sr.uflags, [self.ctd_sk], sr.request_id)
+        sr.datasets[self.ctd_sk].events = self.ctd_events
+        sr.datasets[self.nut_sk].events = self.nut_events
+        sr.datasets[self.ctd_sk]._insert_dataset(ctdpf_ds)
+        sr.datasets[self.nut_sk]._insert_dataset(nutnr_ds)
+        return sr
+
+    @requests_mock.mock()
+    def calculate_nut_sr(self, m):
+        """
+        Calculate a dataset and run QARTOD tests on it. Return the dataset for examination.
+        """
+        # Mock calls to EDEX to fetch QARTOD test records
+        # requests_mock.response chokes when trying to serialize QartodTestRecord objects, so pass the serialized JSON
+        # via the 'text' argument for request mocking instead of using the 'json' argument
+        nut_json = "[%s, %s]" % (NUTNR_QARTOD_RECORD_1, NUTNR_QARTOD_RECORD_2)
+        nut_path = self.qartod_find_url + "?sensor=4A-NUTNRA102"
+        m.get(nut_path, text=nut_json)
+        ctd_json = "[%s, %s]" % (CTDBP_QARTOD_RECORD_1, CTDBP_QARTOD_RECORD_2)
+        ctd_path = self.qartod_find_url + "?sensor=2A-CTDPFA107"
+        m.get(ctd_path, text=ctd_json)
+
+        sr = self.create_nut_sr()
+        sr.calculate_derived_products()
+        sr.import_extra_externals()
+        sr.execute_qartod_qc()
+        sr.insert_provenance()
+        return sr
+
+    def test_expand_qartod_record_for_null_subsite(self):
+        record = self.get_modified_ctdbp_record(subsite=None)
+        expected = [CTDBP_QARTOD_RECORD_3]
+        parameters = ['ctdbp_no_seawater_conductivity', 'ctdbp_no_seawater_pressure', 'practical_salinity',
+                      'seawater_temperature']
+        result = qartodTestServiceAPI.expand_qartod_record(record, 'CE02SHBP', 'LJ01D', '06-CTDBPN106',
+                                                           'ctdbp_no_sample', parameters)
```

Comment:
I recommend you make all the repeating literals to define the elements of "parameters" and the arguments the call to populate "result" constants.

Reply:
Refactored parameters assignment, call to expansion function, and assertion into the method check_ctdbp_record_expansion.